impl/src/main/java/org/ehcache/internal/classes/ClassInstanceProvider.java

```diff
@@ -84,25 +84,44 @@ private T newInstance(K alias, ClassInstanceConfiguration<T> config) {
         return null;
       }
     }
-    try {
-      T instance = invokeConstructor(config.getClazz(), config.getArguments());
-      created.add(instance);
-      return instance;
-    } catch (InstantiationException e) {
-      throw new RuntimeException(e);
-    } catch (IllegalAccessException e) {
-      throw new RuntimeException(e);
-    } catch (NoSuchMethodException e) {
-      throw new RuntimeException(e);
-    } catch (InvocationTargetException e) {
-      throw new RuntimeException(e);
+    
+    T instance = null;
+    
+    if(config.getInstance() != null) {
+      instance = config.getInstance();
+    } else {
+      try {
+        instance = invokeConstructor(config.getClazz(), config.getArguments());
+      } catch (InstantiationException e) {
+        throw new RuntimeException(e);
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      } catch (NoSuchMethodException e) {
+        throw new RuntimeException(e);
+      } catch (InvocationTargetException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    AtomicInteger currentCount = providedVsCount.get(instance);
+    if(currentCount == null) {
+      providedVsCount.put(instance, new AtomicInteger(1));
+    } else {
+      currentCount.incrementAndGet();
     }
+    return instance;
   }
 
   protected void releaseInstance(T instance) throws IOException {
-    if(!created.remove(instance)) {
+    if(providedVsCount.containsKey(instance)) {
```

Comment:
For the instances that are actually *created* by this class the count is never gonna be more than 1. And those instances when released will be removed in the if-block below. Now for the user provided instances the count can be more than 1 if the user reuses the same instance across caches. So here we are doing reference counting to handle that case and again the if block below would handle removing the last instance.